typed-racket-lib/typed-racket/private/type-contract.rkt

```diff
@@ -683,7 +716,30 @@
             (fail #:reason (~a "cannot import structure types from"
                                "untyped code"))
             (struct-type/sc null))]
-       [(Struct-Property: s) (struct-property/sc (t->sc s))]
+       [(Struct-Property: s _) (struct-property/sc (t->sc s))]
+       [(Has-Struct-Property: orig-id)
+        ;; we can't call syntax-local-value/immediate in has-struct-property case in parse-type
+        (define-values (a prop-name) (syntax-local-value/immediate orig-id (λ () (values #t orig-id))))
+        (match-define (Struct-Property: _ pred?) (lookup-id-type/lexical prop-name))
+        ;; if original-name is only set when the type is added via require/typed
+        
+        ;; the original-name of `prop-name` is its original referece in the unexpanded program.
+        (define real-prop-var (or (syntax-property prop-name 'original-name) prop-name))
+
+        ;; a property is wrapped so we need its original reference
+        (define real-pred-var (or (syntax-property pred? 'original-name) (syntax-e pred?)))
+        
+        ;; the `pred?` could be provided to a property through require/typed, 
+        ;; so we need to check if it is produced by the property
+        (flat/sc #`(if (not (struct-type-property-predicate-procedure? #,pred? #,real-prop-var))
+                       (raise-arguments-error 'struct-property
+                                              "predicate does not match property"
+                                              "predicate"
+                                              #,pred?
+                                              "property"
+                                              #,real-prop-var)
+                       (flat-named-contract '#,real-pred-var
+                                            #,pred?)))]
```

Comment:
Should this be `(lambda (x) pred?)`? That`s what `Struct: ` uses, and it's important when the contract def comes above the struct def.

I don't have an example for why Struct: needs a lambda on hand, but `require/typed` might be important to make one.